#+author: Carsten Fortmann-Grote (@CFGrote)
#+date: [2025-12-11 Thu] 
#+property: header-args:R :results output :tangle fastani_heatmap.R :noweb yes :eval yes :session rbioinfo

* Introduction
:PROPERTIES:
:ID:       6e6db8c3-8bdf-420d-aeb7-126193a8627e
:END:
This repository contains code and data for two data processing steps referenced in the Microbial Resource Announcement
"Complete genome sequences of four Ochrobactrum pituitosum strains and three Pseudochrobactrum saccharolyticum strains isolated from C. elegans gut microbiomes".

** gyrA extraction
:PROPERTIES:
:ID:       3b720093-281c-4d1c-b13f-9b89cf669dcc
:END:
The R markdown document /gyrA_sequence_extraction.Rmd/ was used to extract gyrA sequences for further analysis in Clustal-Omega.

** fastANI results analysis
:PROPERTIES:
:ID:       e94c230f-c3b9-46cd-8cf3-f19cf75c50aa
:END:
The scripts /filter_ani_data.sh/ and /fastani_heatmap.R/ were used to generate Fig. 1 in the paper

*** Data filtering
:PROPERTIES:
:ID:       35471fb5-7cb3-459d-a6a8-b999e734cacb
:END:
The data produced by running =fastANI= on our sequence data contains 12*11=132 pairwise ANI similarities, corresponding
to our 8 assembled genomes and 4 reference type strains.
Some cleanup steps need to be taken:
- remove strain NV9 because of inferior data quality
- remove redundant type strains P.sac. CCUG33852 and O.pit. BU72 being close to 100% identical to their respective
  counterparts P.sac. DSM25260 and O.pit. CCUG50899.
- abbreviate Ochrobactrum pituitosum as O.pit. and Pseudochrobactrum saccharolyticum as P.sac.
- use isolate identifiers (MPB numbers) instead of sample titles in strain names
|              |          |                                   |
| Sample Title | Isolate  | Organism                          |
|--------------+----------+-----------------------------------|
| IV11         | MPB25535 | Ochrobactrum pituitosum           |
| PH15         | MPB25415 | Ochrobactrum pituitosum           |
| PH4          | MPB25440 | Ochrobactrum pituitosum           |
| NH1          | MPB25478 | Ochrobactrum pituitosum           |
| IV4          | MPB25528 | Pseudochrobactrum saccharolyticum |
| NH11         | MPB25488 | Pseudochrobactrum saccharolyticum |
| PH5          | MPB25441 | Pseudochrobactrum saccharolyticum |


We run the following shell command:
#+begin_src shell :tangle filter_ani_data.sh
  #! /bin/bash
  cat FastANI_Results.csv \
    | grep -v "CCUG33852" \
    | grep -v "BU72" \
    | grep -v "NV9" \
    | sed 's/IV11/MPB25535/g' \
    | sed 's/PH15/MPB25415/g' \
    | sed 's/PH4/MPB25440/g' \
    | sed 's/NH11/MPB25488/g' \
    | sed 's/NH1/MPB25478/g' \
    | sed 's/IV4/MPB25528/g' \
    | sed 's/PH5/MPB25441/g' \
    | sed 's/Pseudochrobactrum M/M/g' \
    | sed 's/O\.pituitosum M/M/g'  \
    | sed 's/O\.pituitosum/O.pit./g' \
    | sed 's/P\.saccharolyticum/P.sac./g' \
    > FastANI_Results_clean.csv
  cat FastANI_Results_clean.csv
#+end_src


*** Heatmap plot
:PROPERTIES:
:ID:       206c49a5-a191-4f73-9f27-bc35a6d1498b
:END:
The following libraries are needed to produce a formatted clustered heatmap from the filtered ANI data:
#+name: libraries
#+begin_src R :results values quiet :session rbioinfo
  # Load libraries
  library('ggcorrheatmap')
  library('reshape2')
  library('ggplot2')
#+end_src

We load the data using =read.csv= treating strings as values, not as factors:
#+begin_src R :session rbioinfo :results values quiet
  # Read data
  data <- read.csv("FastANI_Results_clean.csv", stringsAsFactors = FALSE)
#+end_src


Next, we reshape the data into a wide format using =dcast=.
We need to make the QUERY values the row names and the REFERENCE values the column names
#+begin_src R :session rbioinfo :results values quiet
  # Reshape into square matrix
  data_matrix <- dcast(data, QUERY ~ REFERENCE, value.var = "ANI_ESTIMATE")
#+end_src

Finally, we remove the QUERY column and insert it as rownames
#+begin_src R :session rbioinfo
  # Remove the QUERY column which is now row names 
  heatmap_data <- as.matrix(data_matrix[, -1])  
  
  # Set QUERY as row names
  rownames(heatmap_data) <- data_matrix$QUERY  
#+end_src

#+RESULTS:

The table reports ANI values as percent identity, we'll recast it to a ratio by dividing by 100

#+begin_src R :session rbioinfo :results values quiet
  # Scale to 0 < ANI < 1.
  heatmap_data <- heatmap_data * 0.01
#+end_src

Finally, we plot the heatmap using =ggplot2::gghm=. We cluster rows and columns but do not render the dendogram. Each
cell is labeled by the corresponding ANI value. We plot only the upper left triangle of the symmetric matrix and
remove the diagonal as it does not add any information. Remaining options and the added =theme= layer are for stylistic
elements. The plot is saved in high resolution to pdf and png files.
#+begin_src R :session rbioinfo :results output file link :file ani_heatmap.png
  # Plot heatmap
  gghm(heatmap_data,
       layout='topleft',
       col_scale='YlGnBu',
       bins=5,
       na_remove=TRUE,
       cell_labels=TRUE,
       cluster_rows=TRUE,
       cluster_cols=TRUE,
       show_dend_rows=FALSE,
       show_dend_cols=FALSE,
       cell_label_digits=3,
       legend_order=NA,
       cell_label_size=6,
       ) + theme(axis.text.x = element_text(size=18), axis.text.y = element_text(size=18))
  ggsave('ani_heatmap.pdf', dpi=600)
  ggsave('ani_heatmap.png', dpi=150)
#+end_src

#+RESULTS:
[[file:ani_heatmap.png]]


To reproduce the Fig. 1 in the Announcement paper, run
#+begin_src shell :eval no
  bash filter_ani_data.sh
  Rscript fastani_heatmap.R
#+end_src
